app/utils/graphql.ts

```diff
@@ -0,0 +1,194 @@
+import { useState, useCallback, useEffect, useMemo } from "react";
+import useSWR, { SWRConfiguration } from "swr";
+import {
+  gql,
+  GraphQLClient,
+  request,
+  RequestDocument,
+  Variables,
+} from "graphql-request";
+import { getHasuraRoleName } from "./auth";
+import { MutationVariables, LookupTableDef } from "@/types/types";
+import { useAuth0 } from "@auth0/auth0-react";
+
+const ENDPOINT = process.env.NEXT_PUBLIC_HASURA_ENDPOINT!;
+
+const DEFAULT_SWR_OPTIONS: SWRConfiguration = {
+  /**
+   * Dont refetch when the page/tab is refocused
+   */
+  revalidateOnFocus: false,
+  /**
+   * Dont refetch on network recon
+   */
+  revalidateOnReconnect: false,
+};
+
+/**
+ * Fetcher which interacts with our GraphQL API
+ */
+const fetcher = <T>([query, variables, token, hasuraRoleName]: [
+  RequestDocument,
+  Variables,
+  string,
+  string
+]): Promise<T> =>
+  request({
+    url: ENDPOINT,
+    document: query,
+    variables,
+    requestHeaders: {
+      Authorization: `Bearer ${token}`,
+      "x-hasura-role": hasuraRoleName,
+    },
+  });
+
+/**
+ * Hook which wraps `useSWR` and provides auth
+ */
+export const useQuery = <T>({
+  query,
+  variables,
+  options,
+}: {
+  query: RequestDocument | null;
+  variables?: Variables;
+  options?: SWRConfiguration;
+}) => {
+  const { getIdTokenClaims, user } = useAuth0();
+
+  const fetchWithAuth = async ([query, variables]: [
+    RequestDocument,
+    Variables
+  ]) => {
+    const hasuraRoleName = getHasuraRoleName(user);
+    /**
+     * todo: our Auth0 app is configured to return idTokens, which are
+     * "opaque" proprietary Auth0 tokens, seemingly because this enables us
+     * to interact with the User Management API. as a result, we must
+     * use idToken.__raw to grab the actual JWT. it seems like our setup
+     * may be misconfigured, because accessing .__raw seems like a hack :/
+     *
+     * the typical setup would enable use to use the getAccessTokenSilently()
+     * method, but that doesn't work with the opaque tokens.
+     *
+     * dicussion here: https://community.auth0.com/t/getting-the-jwt-id-token-from-auth0-spa-js/28281/3
+     */
+    const idToken = await getIdTokenClaims();
+    const token = idToken?.__raw || "";
+
+    // todo: <T> passed here is assigned to query results without validation - we need to use a schema validator
+    return fetcher<T>([query, variables, token, hasuraRoleName]);
+  };
+
+  // todo: document falsey query handling
+  const { data, error, isLoading, mutate, isValidating } = useSWR<T>(
+    query ? [query, variables] : null,
+    fetchWithAuth,
+    { ...DEFAULT_SWR_OPTIONS, ...(options || {}) }
+  );
```

Comment:
here's where we call the actual `useSWR` hook—and a few important things to note:
- the first argument is the value that SWR uses as the cache key. so each combination of query + variables gets cached.
- passing `null` to the first argument is like an off switch, and enables [conditional data fetching](https://swr.vercel.app/docs/conditional-fetching#conditional). this comes in super handy when you don't want to start a request until you have, e.g, a user's email address from the auth hook.

Reply:
I think this may have answered my lazy query questions. 👀

Reply:
yeah!